Occupant standby mode control

lymereJ opened the pull request:
The proposed changes apply standby mode controls to space types targeted by ASHRAE 90.1-2019. This implementation uses a simplified approach where controls are only modeled for one hour each day, the occupancy profile for the targeted spaces is left unchanged. The standby mode control require a setup/setback of the thermostat setpoint and shutoff of the ventilation. The shutoff of the ventilation for zones served by unitary systems is done by setting the fan operation to cycle during that hour and for zone served by a VAV box, the minimum damper position is scheduled to a low value during that hour. Additional information regarding about this requirement are provided in Section B2.2 in the Energy Savings Analysis: [ANSI/ASHRAE/IES Standard 90.1-2019 report](https://www.energycodes.gov/sites/default/files/2021-07/20210407_Standard_90.1-2019_Determination_TSD.pdf).

lymereJ commented on lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2016/ashrae_90_1_2016.PumpVariableSpeed.rb:
```diff
@@ -76,7 +76,7 @@ def pump_variable_speed_get_control_type(pump, plant_loop_type, pump_nominal_hp)
             threshold = 150
           else
-            OpenStudio.logFree(OpenStudio::Warn, "Pump flow control requirement missing for heating water pumps in climate zone: #{climate_zone}.")
+            OpenStudio.logFree(OpenStudio::Warn, 'openstudio.ashrae_90_1_2016.PumpVariableSpeed', "Pump flow control requirement missing for heating water pumps in climate zone: #{climate_zone}.")
```
Unrelated but necessary logging changes.

lymereJ commented on lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:
```diff
@@ -0,0 +1,76 @@
+  #
+  # @param space [OpenStudio::model::Space] OpenStudio Space object
+  # @return [Boolean] true if occupancy standby mode is to be modeled, false otherwise
+  def space_is_occupancy_standby_mode(space)
+    # Get space type
+    return false if space.spaceType.empty?
+
+    space_type = space.spaceType.get
+
+    # Get standards space type
+    return false if space_type.standardsSpaceType.empty?
+
+    std_space_type = space_type.standardsSpaceType.get
+
+    # Space with standby mode are determined based
+    # on note H in Std 62.1 and their automatic partial
+    # of full off lighting control requirement in 90.1.
+    # In 90.1-2019/62.1-2016 this comes down to office
+    # spaces (enclosed =< 250 ft2) and conference/meeting
+    # and multipurpose rooms.
+    # Currently standards doesn't excatly use the 90.1
+    # space description so all spaces types that include
+    # office/meeting/conference are flagged as having
+    # occupant standby mode.
+    if std_space_type.downcase.include?('office') || std_space_type.downcase.include?('meeting') || std_space_type.downcase.include?('conference')
+      return true
+    end
+
+    return false
+  end
```
New function that identifies spaces required to have occupant standby controls.

lymereJ commented on lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:
```diff
@@ -0,0 +1,76 @@
+  #
+  # @param space [OpenStudio::model::Space] OpenStudio Space object
+  # @return [Boolean] true if success
+  def space_occupancy_standby_mode(thermostat)
+    htg_sch = thermostat.getHeatingSchedule.get
+    clg_sch = thermostat.getCoolingSchedule.get
+
+    # Setback heating schedule
+    # Setback is 1 deg. F per code requirement
+    # Time of the day is arbitrary lack of dynamic occupant modeling
+    setup = 1 # deg. F
+    htg_sch_mod = { '12' => -1 * (OpenStudio.convert(1, 'F', 'C').get - OpenStudio.convert(1 - setup, 'F', 'C').get) }
+    htg_sch_name = "#{htg_sch.name} - occupant standby mode"
+    htg_sch_old = thermostat.model.getScheduleRulesetByName(htg_sch_name)
+    if htg_sch_old.empty?
+      htg_sch_offset = model_offset_schedule_value(htg_sch, htg_sch_mod)
+      htg_sch_offset.setName(htg_sch_name)
+      thermostat.setHeatingSchedule(htg_sch_offset)
+    else
+      thermostat.setHeatingSchedule(htg_sch_old.get)
+    end
+
+    # Setup cooling schedule
+    # Setup is 1 deg. F per code requirement
+    # Time of the day is arbitrary lack of dynamic occupant modeling
+    setback = 1 # deg. F
+    clg_sch_mod = { '12' => (OpenStudio.convert(1, 'F', 'C').get - OpenStudio.convert(1 - setback, 'F', 'C').get) }
+    clg_sch_name = "#{clg_sch.name} - occupant standby mode"
+    clg_sch_old = thermostat.model.getScheduleRulesetByName(clg_sch_name)
+    if clg_sch_old.empty?
+      clg_sch_offset = model_offset_schedule_value(clg_sch, clg_sch_mod)
+      clg_sch_offset.setName(clg_sch_name)
+      thermostat.setCoolingSchedule(clg_sch_offset)
+    else
+      thermostat.setCoolingSchedule(clg_sch_old.get)
+    end
+
+    return true
+  end
```
New function that adjust the thermostat of zone with spaces required to have occupant standby mode controls.

lymereJ commented on lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
+  # @param time_offset_hash [Hash] Hash providing time (key) and schedule value offset (values)
+  # @param schedule [OpenStudio::model::ScheduleRuleset] OpenStudio schedule object
+  # @return [OpenStudio::model::ScheduleRuleset] Modified OpenStudio schedule object
+  def model_offset_schedule_value(schedule, time_offset_hash)
+    offset_sch = schedule.clone(schedule.model).to_ScheduleRuleset.get
+
+    # Get day schedule
+    day_schedules = []
+    default_day_schedule = offset_sch.defaultDaySchedule
+    day_schedules << default_day_schedule
+    offset_sch.scheduleRules.each do |rule|
+      day_schedules << rule.daySchedule
+    end
+
+    # Offset schedule values
+    day_schedules.each do |day_schedule|
+      (0..23).each do |hr|
+        t = OpenStudio::Time.new(0, hr, 0, 0)
+
+        # Get schedule value
+        value = day_schedule.getValue(t)
+
+        # Offset schedule value
+        if time_offset_hash.key?(hr.to_s)
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, value + time_offset_hash[hr.to_s])
+        end
+      end
+    end
+
+    offset_sch.setName("#{schedule.name} - offset")
+    return offset_sch
+  end
```
New method to offset schedule values using a hash that associate the time of the day to the offset value.

lymereJ commented on lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
+  # @param time_offset_hash [Hash] Hash providing time (key) and schedule value offset (values)
+  # @param schedule [OpenStudio::model::ScheduleRuleset] OpenStudio schedule object
+  # @return [OpenStudio::model::ScheduleRuleset] Modified OpenStudio schedule object
+  def model_set_schedule_value(schedule, time_value_hash)
+    return nil unless schedule.to_ScheduleRuleset.is_initialized
+
+    new_sch = schedule.clone(schedule.model).to_ScheduleRuleset.get
+
+    # Get day schedule
+    day_schedules = []
+    default_day_schedule = new_sch.defaultDaySchedule
+    day_schedules << default_day_schedule
+    new_sch.scheduleRules.each do |rule|
+      day_schedules << rule.daySchedule
+    end
+
+    # Offset schedule values
+    day_schedules.each do |day_schedule|
+      (0..23).each do |hr|
+        t = OpenStudio::Time.new(0, hr, 0, 0)
+
+        # Get schedule value
+        value = day_schedule.getValue(t)
+
+        # Offset schedule value
+        if time_value_hash.key?(hr.to_s)
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, time_value_hash[hr.to_s])
+        end
+      end
+    end
+
+    new_sch.setName("#{schedule.name} - adjusted")
+    return new_sch
+  end
```
New method that sets a values of a schedule. Time are value are provided into a hash.

lymereJ commented on lib/openstudio-standards/standards/Standards.ZoneHVACComponent.rb:
```diff
@@ -171,6 +171,35 @@ def zone_hvac_component_apply_standard_controls(zone_hvac_component)
       zone_hvac_component_apply_vestibule_heating_control(zone_hvac_component)
     end
 
+    # Convert to objects
+    zone_hvac_component = if zone_hvac_component.to_ZoneHVACFourPipeFanCoil.is_initialized
+                            zone_hvac_component.to_ZoneHVACFourPipeFanCoil.get
+                          elsif zone_hvac_component.to_ZoneHVACPackagedTerminalAirConditioner.is_initialized
+                            zone_hvac_component.to_ZoneHVACPackagedTerminalAirConditioner.get
+                          elsif zone_hvac_component.to_ZoneHVACPackagedTerminalHeatPump.is_initialized
+                            zone_hvac_component.to_ZoneHVACPackagedTerminalHeatPump.get
+                          end
+
+    # Do nothing for other types of zone HVAC equipment
+    if zone_hvac_component.nil?
+      return true
+    end
+
+    # Standby mode occupancy control
+    return true unless zone_hvac_component.thermalZone.empty?
+
+    thermal_zone = zone_hvac_component.thermalZone.get
+
+    standby_mode_spaces = []
+    thermal_zone.spaces.sort.each do |space|
+      if space_is_occupancy_standby_mode(space)
+        standby_mode_spaces << space
+      end
+    end
+    if !standby_mode_spaces.empty?
+      zone_hvac_model_standby_mode_occupancy_control(zone_hvac_component)
+    end
+
```
Handles occupant standby control for zonal systems.

lymereJ commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.AirLoopHVAC.rb:
```diff
@@ -453,4 +453,67 @@ def air_loop_hvac_energy_recovery_ventilator_flow_limit(air_loop_hvac, climate_z
+      unitary_system.setSupplyAirFanOperatingModeSchedule(new_sch) unless new_sch.nil?
+    else
+      # Get thermal zones associated with spaces having standby mode occupancy requirements
+      standby_mode_zones = []
+      standby_mode_spaces.sort.each do |space|
+        standby_mode_zones << space.thermalZone.get
+      end
+      # Schedule the MDP of terminals to a low value during occupant standby mode
+      # The intent is to reduce ventilation while still allowing the terminal to
+      # meet loads
+      standby_mode_zones.each do |zone|
+        air_terminal = zone.airLoopHVACTerminal
+        if air_terminal.is_initialized
+          air_terminal = air_terminal.get
+          if air_terminal.to_AirTerminalSingleDuctVAVReheat.is_initialized
+            air_terminal = air_terminal.to_AirTerminalSingleDuctVAVReheat.get
+            if air_terminal.zoneMinimumAirFlowInputMethod == 'Constant' || air_terminal.zoneMinimumAirFlowInputMethod == 'FixedFlow'
+              if air_terminal.zoneMinimumAirFlowInputMethod == 'FixedFlow'
+                mdp_org = air_terminal.fixedMinimumAirFlowRate.get / air_terminal.autosizedMaximumAirFlowRate.get
+                air_terminal.setFixedMinimumAirFlowRate(0)
+              else
+                mdp_org = air_terminal.constantMinimumAirFlowFraction.get
+                air_terminal.setConstantMinimumAirFlowFraction(0)
+              end
+              air_terminal.setZoneMinimumAirFlowInputMethod('Scheduled')
+              air_terminal.setMinimumAirFlowFractionSchedule(model_set_schedule_value(model_add_constant_schedule_ruleset(air_loop_hvac.model, mdp_org, name = "#{air_terminal.name} - MDP", sch_type_limit: 'Fraction'), '12' => 0.1))
+            elsif air_terminal.zoneMinimumAirFlowInputMethod == 'Scheduled'
+              air_terminal.setMinimumAirFlowFractionSchedule(model_set_schedule_value(air_terminal.minimumAirFlowFractionSchedule.get, '12' => 0.1))
+            else
+              OpenStudio.logFree(OpenStudio::Warn, 'openstudio.ashrae_90_1_2019.AirLoopHVAC', "The air terminal associated with #{zone.name} uses a zone minimum air flow input method that is currently not supported so occupant standby controls were not modeled.")
+            end
+          else
+            OpenStudio.logFree(OpenStudio::Warn, 'openstudio.ashrae_90_1_2019.AirLoopHVAC', "The air terminal associated with #{zone.name} isn't of the SingleDuctVAVReheat type so occupant standby controls were not modeled.")
+          end
+        end
+      end
+    end
+
+    return true
+  end
```
New method that sets fan operating mode of unitary system to be cycling during occupant standby mode period, and schedule MDP of VAV boxes to a lower value during occupant standby mode period.

mdahlhausen reviewed (changes requested):
space_occupancy_standby_mode(thermostat) is similar in naming to space_is_occupancy_standby_mode(space).
We've generally used _required? at the end of methods to check if a certain control scheme is required.  Ample examples in AirLoopHVAC.rb.

Also, the documentation is confusing in [Prototype.Model.rb](lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb):
# Modify thermostat schedule to account for a thermostat setback/up
#
# @param space [OpenStudio::model::Space] OpenStudio Space object
# @return [Boolean] true if success
def space_occupancy_standby_mode(thermostat)
  return true
end

@param space should be @param thermostat.  And why does it default to true instead of false?

Some other thoughts - the setback time is arbitrary, but could be knowable in the model from the schedule hours of operation additional properties object, or from the methods that generate a boolean zone occupancy.  Note sure how to address that.  But I suspect this may cause some issues with more generic models applying the 90.1-2019 logic.

lymereJ commented:
> space_occupancy_standby_mode(thermostat) is similar in naming to space_is_occupancy_standby_mode(space).
> We've generally used _required? at the end of methods to check if a certain control scheme is required. Ample examples in AirLoopHVAC.rb.

That makes sense. Addressed in the latest commit.

> Also, the documentation is confusing in [Prototype.Model.rb](lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb):
> 
> ```
> # Modify thermostat schedule to account for a thermostat setback/up
> #
> # @param space [OpenStudio::model::Space] OpenStudio Space object
> # @return [Boolean] true if success
> def space_occupancy_standby_mode(thermostat)
>   return true
> end
> ```
> 
> @param space should be @param thermostat. And why does it default to true instead of false?

Good catch, I changed my approach and forgot to change the docs.

> Some other thoughts - the setback time is arbitrary, but could be knowable in the model from the schedule hours of operation additional properties object, or from the methods that generate a boolean zone occupancy. Note sure how to address that. But I suspect this may cause some issues with more generic models applying the 90.1-2019 logic.

I agree, do you think that it would make sense to open an issue and keep this PR as-is for now?

mdahlhausen reviewed (changes requested):
Changes look good.

I think these methods should be generalized to take a zone or space occupancy schedule to create the offset:
`model_offset_schedule_value(schedule, time_offset_hash)`
`space_occupancy_standby_mode(thermostat)`

We can hold off on the merge for now.

lymereJ commented:
@mdahlhausen  - `model_offset_schedule_value(schedule, time_offset_hash)` was meant to be a helper function that could be used by any other function. Refactoring to take a zone of space occupancy schedule seems like it would reduce its scope of applicability. Don't you think?

If we want to generalize `space_occupancy_standby_mode(thermostat)` we also need to generalize `zone_hvac_model_standby_mode_occupancy_control(zone_hvac_component)`. When you say hold off on the merge, do you mean as part of the next release?

mdahlhausen commented on lib/openstudio-standards/prototypes/ashrae_90_1/ashrae_90_1_2019/ashrae_90_1_2019.Space.rb:
```diff
@@ -0,0 +1,76 @@
+    # Setback is 1 deg. F per code requirement
+    # Time of the day is arbitrary lack of dynamic occupant modeling
+    setup = 1 # deg. F
+    htg_sch_mod = { '12' => -1 * (OpenStudio.convert(1, 'F', 'C').get - OpenStudio.convert(1 - setup, 'F', 'C').get) }
```
use Ranking and Kelvin, not Farenheit and Celsius

lymereJ replied:
Addressed in latest commit.

mdahlhausen commented on lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
+        if time_offset_hash.key?(hr.to_s)
+          day_schedule.addValue(t, value)
+          t_p_1 = OpenStudio::Time.new(0, hr + 1, 0, 0)
+          day_schedule.addValue(t_p_1, value + time_offset_hash[hr.to_s])
```
does this override?

lymereJ replied:
No, because it sets it for the next hour.

mdahlhausen commented on lib/openstudio-standards/prototypes/common/objects/Prototype.Model.rb:
```diff
@@ -2546,4 +2557,94 @@ def model_add_transfer_air(model)
+    # Offset schedule values
+    day_schedules.each do |day_schedule|
+      (0..23).each do |hr|
+        t = OpenStudio::Time.new(0, hr, 0, 0)
```
guard clause here?

mdahlhausen replied:
next if hr not in time_offset_hash

lymereJ replied:
Addressed in latest commit.
